File: libs/carbon/carbon.cpp
#include "pxt.h"
#include "CarbonService.h"

enum class GenerationMixType{
    gas = 0,
    coal = 1, 
    solar = 2, 
    wind = 3, 
    nuclear = 4, 
    hydro = 5, 
    biomas = 6, 
    imports = 7, 
    other = 8,
};

enum class CarbonIndex{
    very_low = 0,
    low = 1, 
    moderate = 2, 
    high = 3, 
    very_high = 4, 
};

static const char* const generationTypes[] = {"gas","coal","solar","wind","nuclear","hydro","biomas","imports","other"}; 
static const char* const carbonIndexValueTypes[] = {"very_low","low","moderate","high","very_high"}; 

/**
* share
* use
*/

CarbonService CarbonService(uBit.radio);

//% color=#000000 weight=100 icon="\uf216" block="carbon"
namespace carbon {

    int init() {
        return enableRadio();
    }

    int stringToNumber(ManagedString input)
    {
        ManagedString resultString = "";
        int result = 0;

        int l = input.length();
        for(int x=0; x < l; ++x)
        {
            if(input.charAt(x) != '\n' && input.charAt(x) != '\r')
            {
                if(isdigit(input.charAt(x)) || input.charAt(x) == '.')
                {
                    resultString = resultString + input.charAt(x);
                }
                else
                {
                    result = -1;
                    break;
                }
            }
        }
        if(result > -1)
        {
            float f = atof(resultString.toCharArray());
            result = round(f);
        }
        return result;
    }

    //%
    int queryCarbonValue()
    {
        init();

        ManagedString reply = CarbonService.getCarbonValue("value");
        int value = stringToNumber(reply);
        return value;
    }

    //Convert the carbon index from string to int
    // 'very low' 'low' 'moderate' 'high' 'very high'
    //  0          1     2          3      4
    //%
    int queryCarbonIndex()
    {
        init();

        ManagedString reply = CarbonService.getCarbonIndex("index");

        int replyValue = -1;

        for(int loop=0; loop < 5; ++ loop)
        {
            //Compare the response text with the items in the array
            //if we get a match, the loop value is the correct number to return
            if(strcmp(carbonIndexValueTypes[loop], reply.toCharArray()) == 0)
            {
                replyValue = loop;
                break;
            }
        }
        return replyValue;
    }

    //%
    StringData* queryCarbonIndexText()
    {
        init();

        ManagedString reply = CarbonService.getCarbonIndex("index");
        return reply.leakData();
    }

    //%
    int queryCarbonGenerationMix(GenerationMixType mix)
    {
        init();

        //Check the input value bounds
        if(mix < GenerationMixType::gas)
        {
            mix = GenerationMixType::gas;
        }

        if(mix > GenerationMixType::other)
        {
            mix = GenerationMixType::other;
        }
               
        ManagedString command = "genmix/";
        command = command + generationTypes[(int)mix];
        
        ManagedString reply = CarbonService.getCarbonGenerationMix(command);
        int value = stringToNumber(reply);
        return value;
    }


    //%
    int setGenerationMixType(GenerationMixType GenerationMixType)
    {
        return (int)GenerationMixType;
    }

    //%
    int setCarbonIndexValueType(CarbonIndex index)
    {
        return (int)index;
    }

    
};

File: libs/carbon/carbon.ts
//% color=#004d00 weight=90 icon="\uf3fd" block="carbon"
namespace carbon {


    //% blockId=carbon_setGenerationMixType block="generation mix type %mixType" weight=75
    //% shim=carbon::setGenerationMixType
    export function setGenerationMixType(mixType:GenerationMixType): number {
        return mixType;
    }

    //% blockId=carbon_setCarbonIndexValueType block="carbon index level %index" weight=75
    //% shim=carbon::setCarbonIndexValueType
    export function setCarbonIndexValueType(index:CarbonIndex): number {
        return index;
    }

    //% blockId=carbon_queryCarbonValue block="get carbon value" weight=74
    //% shim=carbon::queryCarbonValue
    export function queryCarbonValue(): number {
        return Math.random(200) + 120;
    }

    //% blockId=carbon_queryCarbonIndex block="get carbon index level" weight=73
    //% shim=carbon::queryCarbonIndex
    export function queryCarbonIndex(): number {
        return Math.random(200) + 120;
    }

    //% blockId=carbon_queryCarbonIndexText block="get carbon index as text" weight=72
    //% shim=carbon::queryCarbonIndexText
    export function queryCarbonIndexText(): string {
        let n = Math.random(5);
        
        switch(n)
        {
            case 0:
                return "very low";
                break;
            case 1:
                return "low";
                break;
            case 2:
                return "moderate";
                break;
            case 3:
                return "high";
                break;
            case 4:
                return "very high";
                break;

        }
        return "moderate";
    }

    //% blockId=carbon_queryCarbonGenerationMix block="get carbon generation mix|for type %mix" weight=71
    //% shim=carbon::queryCarbonGenerationMix
    export function queryCarbonGenerationMix(mix : GenerationMixType): number 
    {
        if(mix == GenerationMixType.biomas){
            return Math.random(4) + 4;
        }
        if(mix == GenerationMixType.coal){
            return Math.random(3) + 1;
        }
        if(mix == GenerationMixType.imports){
            return Math.random(5) + 4;
        }
        if(mix == GenerationMixType.gas){
            return Math.random(43) + 20;
        }
        if(mix == GenerationMixType.nuclear){
            return Math.random(11) + 10;
        }
        if(mix == GenerationMixType.hydro){
            return Math.random(3) + 1;
        }
        if(mix == GenerationMixType.solar){
            return Math.random(11) + 10;
        }
        if(mix == GenerationMixType.wind){
            return Math.random(8) + 2;
        }
        return 0;
    }
}